src/v1/parser.rs

```diff
@@ -0,0 +1,814 @@
+    type Item = Result<Entry, ParseError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut buf = vec!();
+        let is_last = itry!(read_line(self.reader.by_ref(), &mut buf));
+        let row = &buf[..];
+        self.current_row_num += 1;
+        let entry = itry!(Entry::parse(
+                row, &self.current_dir, self.hash_type, self.block_size)
+            .context(self.current_row_num));
+        match entry {
+            None => {
+                let _footer = itry!(Footer::parse(row, self.hash_type)
+                    .context(self.current_row_num));
+                if is_last {
+                    return Some(Err(ParseError::Parse(
+                        format!("Footer must be ended by a newline"),
+                        self.current_row_num)));
+                }
+                let mut test_buf = [0; 1];
+                if itry!(self.reader.read(&mut test_buf)) != 0 {
+                    return Some(Err(ParseError::Parse(
+                        format!("Found extra lines after the footer"),
+                        self.current_row_num)));
+                }
+                None
+            },
+            Some(entry) => {
+                if let Entry::Dir(ref dir_path) = entry {
+                    self.current_dir = dir_path.clone();
+                }
+                Some(Ok(entry))
+            },
+        }
+    }
+}
+
+fn read_line<R: BufRead>(reader: &mut R, mut buf: &mut Vec<u8>)
+    -> Result<bool, io::Error>
+{
+    let _ = reader.read_until(b'\n', &mut buf)?;
+    let is_last = if buf.ends_with(b"\n") {
+        buf.pop();
+        false
+    } else {
+        true
+    };
+    Ok(is_last)
+}
+
+fn parse_path<'a>(data: &'a [u8]) -> (Cow<Path>, &'a [u8]) {
+    let (path, tail) = parse_os_str(data);
+    let unescaped_path = match unescape_hex(path) {
+        Cow::Borrowed(p) => Cow::Borrowed(Path::new(p)),
+        Cow::Owned(p) => Cow::Owned(PathBuf::from(&p)),
+    };
+    (unescaped_path, tail)
+ }
+
+fn parse_path_buf<'a>(data: &'a [u8]) -> (PathBuf, &'a [u8]) {
+    let (path, tail) = parse_os_str(data);
+    let unescaped_path = unescape_hex(path);
+    (PathBuf::from(&unescaped_path), tail)
+}
+
+fn parse_os_str<'a>(data: &'a [u8]) -> (&OsStr, &'a [u8]) {
+    let (field, tail) = parse_field(data);
+    (OsStr::from_bytes(field), tail)
+}
+
+fn parse_u64<'a>(data: &'a [u8]) -> Result<(u64, &'a [u8]), ParseRowError> {
+    let (field, tail) = parse_field(data);
+    let v = try!(std::str::from_utf8(field).map_err(|e| {
+        ParseRowError(format!("Cannot parse integer {:?}: {}",
+            String::from_utf8_lossy(field).into_owned(), e))}));
+
+    let v = try!(u64::from_str_radix(v, 10).map_err(|e| {
+        ParseRowError(format!("Cannot parse integer {:?}: {}",
+            String::from_utf8_lossy(field).into_owned(), e))}));
+    Ok((v, tail))
+}
+
+fn parse_field<'a>(data: &'a [u8]) -> (&'a [u8], &'a [u8]) {
```

Comment:
Wow! Then I have two points:

1. File is all ASCII you can decode it at any time without loosing anything (have I missed this point earlier?)
2. `slice::Split` does not allocate too. You might have a problem that you can't name a split type because it requires a closure. But I doesn't have to be a problem.